### PR TITLE
xtensa/esp32: Fix cache issue detected by DEBUG_ASSERTION

### PR DESCRIPTION
**Pull request by acassis**
## Summary
xtensa/esp32: Fix cache issue detected by DEBUG_ASSERTION
## Impact
User will not face a DEBUG ASSERTION when using PSRAM
## Testing
ESP32

NOTE: The CI will fail because are some pre-existing Mixed Case in the esp32_spiflash.c file

### CODE REVIEW

File: arch/xtensa/src/esp32/esp32_spicache.h [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+ *
+ * Description:
+ *   Enable cache for SPI FLASH on a CPU
+ *
+ * Input Parameters:
+ *   - cpu: ID of CPU to enable cache
+ *   - state: the cache reg state to be set
+ *
+ * Returned Value:
+ *   None (the return will be over *state)
```
Comment (gustavonihei) [inline arch/xtensa/src/esp32/esp32_spicache.h:79]:
```suggestion
 *   None
```
Here `state` is just an input parameter

File: arch/xtensa/src/esp32/esp32_spiram.c [hunk 2]
```diff
@@ -97,6 +99,141 @@ size_t __attribute__((weak)) esp_himem_reserved_area_size(void)
...
+                     ((vaddr & (0x3fffff >> mask_s)) >> shift);
+        }
+      else
+        {
+          return MMU_SET_VADDR_OUT_RANGE;
+        }
+    }
+
+  /* The MMU registers are implemented in such a way that lookups from the
+   * cache subsystem may collide withCPU access to the MMU registers. We use
```
Comment (gustavonihei) [inline arch/xtensa/src/esp32/esp32_spiram.c:189]:
```suggestion
   * cache subsystem may collide with CPU access to the MMU registers. We use
```

File: arch/xtensa/src/esp32/esp32_spiram.c [hunk 3]
```diff
@@ -97,6 +99,141 @@ size_t __attribute__((weak)) esp_himem_reserved_area_size(void)
...
+   */
+
+  flags = spin_lock_irqsave();
+
+#ifdef CONFIG_SMP
+  spi_disable_cache(0, &statecpu0);
+  spi_disable_cache(1, &statecpu1);
+#else
+  spi_disable_cache(0, &statecpu0);
+#endif
```
Comment (gustavonihei) [inline arch/xtensa/src/esp32/esp32_spiram.c:200]:
```suggestion
  spi_disable_cache(0, &statecpu0);
 
#ifdef CONFIG_SMP
  spi_disable_cache(1, &statecpu1);
#endif
```
nit: I think that avoiding the `#else` statement makes the code cleaner.

File: arch/xtensa/src/esp32/esp32_spiram.c [hunk 4]
```diff
@@ -97,6 +99,141 @@ size_t __attribute__((weak)) esp_himem_reserved_area_size(void)
...
+      regval |= mask_s;
+      putreg32(regval, DPORT_APP_CACHE_CTRL1_REG);
+    }
+
+#ifdef CONFIG_SMP
+  spi_enable_cache(0, statecpu0);
+  spi_enable_cache(1, statecpu1);
+#else
+  spi_enable_cache(0, statecpu0);
+#endif
```
Comment (gustavonihei) [inline arch/xtensa/src/esp32/esp32_spiram.c:231]:
```suggestion
  spi_enable_cache(0, statecpu0);
#ifdef CONFIG_SMP
  spi_enable_cache(1, statecpu1);
#endif
```
nit: One less `#else` statement

**Comment by yamt** [comment]
which debug assertion are you talking about?
does this basically replace the rom version of cache_sram_mmu_set with another implementation?
how does it fix the assertion?

File: arch/xtensa/src/esp32/esp32_spicache.h (hunk 1 above)
Comment (acassis) [reply arch/xtensa/src/esp32/esp32_spicache.h:79]: hi @gustavonihei yes, but the state will be saved in the pointer *state and will be used by the code that called the function

File: arch/xtensa/src/esp32/esp32_spiram.c (hunk 3 above)
Comment (acassis) [reply arch/xtensa/src/esp32/esp32_spiram.c:200]: Good catch! Thank you!

**Comment by acassis** [comment]
> which debug assertion are you talking about?
> does this basically replace the rom version of cache_sram_mmu_set with another implementation?
> how does it fix the assertion?

Hi @yamt the ROM implementation has an issue, you can find more info about it here:
esp-idf/components/esp32/cache_sram_mmu.c

The DEBUGASSERT issue was happening in the function set_bank() at esp32_himem.c because the ROM fuction didn't return 0.

**Comment by yamt** [comment]
> > which debug assertion are you talking about?
> > does this basically replace the rom version of cache_sram_mmu_set with another implementation?
> > how does it fix the assertion?
> 
> Hi @yamt the ROM implementation has an issue, you can find more info about it here:
> esp-idf/components/esp32/cache_sram_mmu.c
> 
> The DEBUGASSERT issue was happening in the function set_bank() at esp32_himem.c because the ROM fuction didn't return 0.

ok. thank you for explanation.

**Comment by yamt** [comment]
maybe you should fix "Blank line contains whitespace" nxstyle complaint. otherwise, LGTM.

**Comment by acassis** [comment]
> maybe you should fix "Blank line contains whitespace" nxstyle complaint. otherwise, LGTM.

Thank you @yamt ! Fixed now!

Please merge it after the CI completes

**Review by Ouss4 — APPROVED** [review]

**Pull request merged**